_posts/2020-08-11-relay-pagingation-containers-demystified.markdown

```diff
@@ -0,0 +1,234 @@
+      name
+      id
+      artworksConnection (first: $count, after: $cursor) @connection(key: "Artist_artworks") {
+        edges {
+          node {
+            title # Now fetch all the artwork data
+            id
+          }
+        }
+      }
+    }
+  `
+}, {
+  direction: "forward",
+  getConnectionFromProps(props) {
+    return props.artist.artworks
+  },
+  getFragmentVariables(prevVars, count) {
+    return {
+      ...prevVars,
+      count,
+    }
+  },
+  getVariables(props, { count, cursor }, fragmentVariables) {
+    return {
+      id: props.artist.id,
+      count,
+      cursor,
+    }
+  },
+  query: graphql` # Here is the query to fetch any specific page
+    query ArtistArtworksQuery(
+      $id: ID!
+      $count: Int!
+      $cursor: String) {
+      artist(id: $id) {
+        ...Artist_artist @arguments(
+          count: $count
+          cursor: $cursor
+        )
+      }
+    }
+  `
+})
+```
+
+Wow, that's a lot! I don't want to get too bogged down in details, so let's break this apart at a high level:
+
+- We changed the React component to show a list of artworks and include a button to load the next page.
+- We changed from `createFragmentContainer` to using `createPaginationContainer`.
+- We added GraphQL fragment variables for `count` and `cursor` to be passed through to the new
+  `artworksConnection`, which we also added.
+- Finally, we added a whole new configuration parameter to `createPaginationContainer`.
+
+This last bit is the part where I see the most frustration. Hopefully what follows will clear things up.
+
+I like to always start by [reading the docs](https://relay.dev/docs/en/pagination-container). The `direction` key
+is the direction that we paginate through, either `"forward"` or `"backward"`. `getConnectionFromProps` is a
+function that returns the GraphQL connection, in case the query has more than one. And `query` is used to fetch any
+specific page of results.
+
+Those all makes sense to me, but then we arrive at the real gotchas: `getFragmentVariables` and `getVariables`. The
+docs are helpful, but only if you understand
+[the internals of how Relay works](https://relay.dev/docs/en/runtime-architecture.html). Relay has a sophisticated
+architecture that delivers some really well-performing code, but its abstractions sometimes
+"[leak](https://en.wikipedia.org/wiki/Leaky_abstraction)" and you have to deal with implementation details that
+Relay doesn't really want you to even be aware of (in this case, the Relay store).
+
+So what are these two functions? Let's return to the docs:
+
+- `getFragmentVariables` is used when re-rendering the component, to retrieve the previously-fetch GraphQL response
+  for a certain set of variables.
+- `getVariables` is used when actually fetching another page, and its return value is given to the `query`.
+
+I think of `getFragmentVariables` as kind of cache key for lookup in Relay's internal store. Our implementation of
```

Comment:
**suggestion:** `as kind of cache key for lookup` reads a little weird to me. Does `as kind of _a_ cache key` sound better?